src/cli/subcommands/archive_cmd.rs

```diff
@@ -0,0 +1,208 @@
+            let child = tipset.take()?;
+            *tipset = Tipset::load(&store, child.parents()).ok().flatten();
+            Some(child)
+        });
+
+        let windowed = (std::iter::once(root).chain(tipsets)).tuple_windows();
+
+        let mut network: String = "unknown".into();
+        let mut lowest_stateroot_epoch = root_epoch;
+        let mut lowest_message_epoch = root_epoch;
+
+        let iter = if progress {
+            itertools::Either::Left(windowed.progress_count(root_epoch as u64))
+        } else {
+            itertools::Either::Right(windowed)
+        };
+
+        for (parent, tipset) in iter {
+            if tipset.epoch() >= parent.epoch() && parent.epoch() != root_epoch {
+                bail!("Broken invariant: non-sequential epochs");
+            }
+
+            if tipset.epoch() < 0 {
+                bail!("Broken invariant: tipset with negative epoch");
+            }
+
+            // Update the lowest-stateroot-epoch only if our parent also has a
+            // state-root. The genesis state-root is usually available but we're
+            // not interested in that.
+            if lowest_stateroot_epoch == parent.epoch() && store.has(tipset.parent_state())? {
+                lowest_stateroot_epoch = tipset.epoch();
+            }
+            if lowest_message_epoch == parent.epoch()
+                && store.has(tipset.min_ticket_block().messages())?
+            {
+                lowest_message_epoch = tipset.epoch();
+            }
+
+            if tipset.epoch() == 0 {
+                if tipset.min_ticket_block().cid().to_string() == calibnet::GENESIS_CID {
+                    network = "calibnet".into();
+                } else if tipset.min_ticket_block().cid().to_string() == mainnet::GENESIS_CID {
+                    network = "mainnet".into();
+                }
+            }
+
+            // If we've already found the lowest-stateroot-epoch and
+            // lowest-message-epoch then we can skip scanning the rest of the
+            // archive when we find a checkpoint.
+            let may_skip =
+                lowest_stateroot_epoch != tipset.epoch() && lowest_message_epoch != tipset.epoch();
+            if may_skip {
+                if let Some(genesis_keys) =
+                    crate::chain::store::index::checkpoint_tipsets::genesis_from_checkpoint_tipset(
+                        tipset.key(),
+                    )
+                {
+                    let genesis_cid = genesis_keys.cids[0];
+                    if genesis_cid.to_string() == calibnet::GENESIS_CID {
+                        network = "calibnet".into();
+                    } else if genesis_cid.to_string() == mainnet::GENESIS_CID {
+                        network = "mainnet".into();
+                    }
+                    break;
+                }
+            }
+        }
+
+        Ok(ArchiveInfo {
+            variant: "CARv1".into(),
```

Comment:
I wonder if this variant info can be made available within car archive; wouldn't need to hardcode it then if we want to support versioning in future

Reply:
Determining whether the content of a file is in the CARv1 or CARv2 format is up to us. The formats are not compatible and have to be treated differently.